src/main/java/gregtech/common/covers/filter/BaseFilterContainer.java

```diff
@@ -0,0 +1,170 @@
+        if (hasFilter()) currentItemFilter.setMaxTransferSize(this.maxTransferSize);
+    }
+
+    public boolean hasFilter() {
+        return currentItemFilter != null;
+    }
+
+    public T getFilter() {
+        return currentItemFilter;
+    }
+
+    public final void setFilter(T newFilter) {
+        this.currentItemFilter = newFilter;
+        if (currentItemFilter != null) {
+            currentItemFilter.setDirtyNotifiable(dirtyNotifiable);
+        }
+        if (onFilterInstanceChange != null) {
+            this.onFilterInstanceChange.run();
+        }
+    }
+
+    public void onFilterInstanceChange() {
+        dirtyNotifiable.markAsDirty();
+    }
+
+    public void setOnFilterInstanceChange(Runnable onFilterInstanceChange) {
+        this.onFilterInstanceChange = onFilterInstanceChange;
+    }
+
+    public boolean showGlobalTransferLimitSlider() {
+        return getMaxTransferSize() > 0 &&
+                (isBlacklistFilter() || !hasFilter() || currentItemFilter.showGlobalTransferLimitSlider());
+    }
+
+    public int getTransferLimit(int slotIndex) {
+        if (isBlacklistFilter() || currentItemFilter == null) {
+            return getTransferSize();
+        }
+        return currentItemFilter.getTransferLimit(slotIndex, getTransferSize());
+    }
+
+    public boolean test(R toTest) {
+        return !hasFilter() || getFilter().test(toTest);
+    }
+
+    public MatchResult<R> match(R toMatch) {
+        if (!hasFilter())
+            return new MatchResult<>(true, toMatch, -1);
+
+        return getFilter().match(toMatch);
+    }
+
+    public int getTransferLimit(R stack) {
+        if (isBlacklistFilter() || currentItemFilter == null) {
+            return getTransferSize();
+        }
+        return currentItemFilter.getTransferLimit(stack, getTransferSize());
+    }
+
+    protected abstract void onFilterSlotChange(boolean notify);
+
+    public int getTransferSize() {
+        if (!showGlobalTransferLimitSlider()) {
+            return getMaxTransferSize();
+        }
+        return this.transferSize;
+    }
+
+    public void setTransferSize(int transferSize) {
+        this.transferSize = MathHelper.clamp(transferSize, 1, getMaxTransferSize());
+        onFilterInstanceChange();
+    }
+
+    @Deprecated
+    public void adjustTransferStackSize(int amount) {
+        setTransferSize(transferSize + amount);
+    }
+
+    public void setBlacklistFilter(boolean blacklistFilter) {
+        if (hasFilter()) getFilter().setBlacklistFilter(blacklistFilter);
+        onFilterInstanceChange();
+    }
+
+    public boolean isBlacklistFilter() {
+        return hasFilter() && getFilter().isBlacklistFilter();
+    }
+
+    /** Uses Cleanroom MUI */
+    public abstract IWidget initUI(ModularPanel main, GuiSyncManager manager);
+
+    public void writeInitialSyncData(PacketBuffer packetBuffer) {
+        packetBuffer.writeItemStack(this.filterInventory.getStackInSlot(0));
```

Comment:
Initial Sync should be the entire filter Inventory, not just the first slot. I had issues with items in filters not in slot 1 disappearing after loading back into the world.

Reply:
`filterInventory` is the handler for the slot that you place the filter in, it only has one slot to serialize.

Reply:
HM, if that is the case I don't know what would be causing my filter saving issue

Reply:
i know that if you use MUI 2.4.1, there's issues with the ItemStackItemHandler not serializing to NBT in certain situations. MUI 2.4.2 resolves this, but introduces the issue with ArrayIndexOutOfBounds

Reply:
This is with MUI 2.4.2 though, and since it is an item filter, I am not getting the AIOOB

Reply:
hmm, under which circumstance is the item not saved to the filter? i am not able to reproduce the issue.

Reply:
Hm, cannot reproduce now.